Student/Student/Notifications/ActivityStreamViewController.swift

```diff
@@ -0,0 +1,131 @@
+        let name: String?
+        let courseCode: String?
+        let color: UIColor?
+    }
+
+    let env = AppEnvironment.shared
+    lazy var activities = env.subscribe(GetActivities()) { [weak self] in
+        self?.update()
+    }
+
+    lazy var courses = env.subscribe(GetCourses()) { [weak self] in
+        self?.cacheCourses()
+    }
+
+    lazy var colors = env.subscribe(GetCustomColors()) { [weak self] in
+        self?.cacheCourses()
+    }
+
+    var courseCache: [String: Info] = [:]
+
+    static func create() -> ActivityStreamViewController {
+        let vc = loadFromStoryboard()
+        return vc
+    }
+
+    override func viewDidLoad() {
+        setupTableView()
+        refreshData()
+    }
+
+    func setupTableView() {
+        tableView.registerHeaderFooterView(SectionHeaderView.self)
+        let refresh = UIRefreshControl()
+        refresh.addTarget(self, action: #selector(refresh(_:)), for: .valueChanged)
+        tableView?.refreshControl = refresh
+    }
+
+    @objc func refresh(_ control: UIRefreshControl) {
+        control.endRefreshing()
+        refreshData(force: true)
+     }
+
+    func refreshData(force: Bool = false) {
+        courses.exhaust(while: { _ in true })
+        activities.refresh(force: force)
+        colors.refresh(force: force)
+    }
+
+    func update() {
+        tableView.reloadData()
+    }
+
+    func cacheCourses() {
+        if colors.pending || courses.pending { return }
+        courses.forEach { [weak self] in self?.courseCache[ $0.id ] = Info(name: $0.name, courseCode: $0.courseCode, color: $0.color) }
+        update()
+    }
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        activities.count
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: ActivityCell = tableView.dequeue(for: indexPath)
+        if let a = activities[indexPath] { cell.update(a, courseCache: courseCache) }
+        return cell
+    }
+}
+
+extension ActivityStreamViewController {
+    public override func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        if scrollView.isBottomReached() {
+            activities.getNextPage()
+        }
+    }
+}
+
+class ActivityCell: UITableViewCell {
+    @IBOutlet weak var titleLabel: DynamicLabel!
+    @IBOutlet weak var subTitleLabel: DynamicLabel!
+    @IBOutlet weak var icon: UIImageView!
+    @IBOutlet weak var courseCode: DynamicLabel!
+
+    func update(_ activity: Activity, courseCache: [String: ActivityStreamViewController.Info] ) {
+        if activity.type == .conversation {
+            titleLabel.text = NSLocalizedString("New Message", comment: "")
+        } else {
+            titleLabel.text = activity.title
+        }
+
+        if activity.context?.contextType == .course || activity.type == .discussion, let id = activity.context?.id {
+            courseCode.text = courseCache[id]?.courseCode
```

Comment:
There's plenty of room for a full course name, and that's what I'm showing on To Do. I think we probably want to align. Would you ping Dave & ask which we should do?

Another potential thing to align is To Do puts the item title first, and the context name next, then the date, while this shows context code, item title, and then date. It's possible Doe wants them to look really different, but they seem so similar.